### ISSUE TITLE
Review AuthN/AuthZ doc & identify next steps

### ISSUE THREAD

**Issue by JoeTice** [department-of-veterans-affairs/va.gov-team#61208]
# Description
Review the write-up that Gia created with analysis of approaches to authn/authz for preview environments.

[Document](https://docs.google.com/document/d/1_LnGqXOFWvaAHDRhIKzkX-0VCL4PI72cQRBF5hwSi2c/edit)

# Tasks
- [x] Review document
- [x] Provide commentary / concurrence

# Acceptance Criteria
- [x] Analysis has been reviewed and commented on

**Comment by pjhill** [comment]
Seems like there are two main options --
1. Environment variable / build parameter + GitHub OAuth
2. Feature toggles + allow list

**Comment by pjhill** [comment]
* We want to run our analysis by the architecture working group
* I wonder if we could borrow anything from Platform Console's Keycloak implementation
* We should check in with the Identity team to see if they have thoughts about what we are thinking we need to do

**Comment by holdenhinkle** [comment]
## Next steps:

### GitHub OAuth:
- [ ] Check with Trevor Bosaw, John Bramely, Joe Niquette to ask them what they think about adding a conditional `before_action` to `ApplicationController` that makes sure the user 1) is authed via GitHub OAuth, or 2) redirects the user to GitHub to be authed.

Example code:
```
class ApplicationController < ActionController::API
  before_action :github_oauth_authentication, unless: -> { Rails.env.production? }

  def github_oauth_authentication
    if session[:github_oauth_token].nil?
      redirect_to "https://github.com/login/oauth/authorize?client_id=#{YOUR_CLIENT_ID}&redirect_uri=#{YOUR_REDIRECT_URI}"
    else
      begin
        # Attempt to use the token to make an authenticated request to the GitHub API
        Octokit.auto_paginate = true
        client = Octokit::Client.new(access_token: session[:github_oauth_token])
        client.user # this will raise an error if the token is invalid
      rescue Octokit::Unauthorized => e
        # If the token is invalid, redirect the user to authenticate again
        redirect_to "https://github.com/login/oauth/authorize?client_id=#{YOUR_CLIENT_ID}&redirect_uri=#{YOUR_REDIRECT_URI}"
      end
    end
  end
end
```
- [ ] Also check with Trevor Bosaw, John Bramely, Joe Niquette, Alex Garcia about including JavaScript like the following on each document load that sends a request to the associated vets-api PE instance:

```
window.onload = function() {
  // Check if the user's OAuth token is stored in the session
  if (!sessionStorage.getItem('githubOAuthToken')) {
    // If not, redirect them to GitHub for authentication
    window.location.href = 'https://github.com/login/oauth/authorize?client_id=YOUR_CLIENT_ID&redirect_uri=YOUR_REDIRECT_URI&scope=YOUR_SCOPES';
  } else {
    // If the user is authenticated, then proceed to load the site content
    loadSiteContent();
  }
};
```

- [ ] If they're okay with the above approaches, look into how to automate it (there doesn't appear to be an api for creating/managing GitHub OAuth Apps)

**Comment by holdenhinkle** [comment]
### Flipper/Feature Flags

The major downside to using feature toggles + allow list (does Flipper gem have an allow-list feature) is that each PE that needs auth would need it's own feature flag. There's a chance we'd end up with tons of PE feature flags which isn't ideal. Also, this is not what feature flags were built for.

**Comment by holdenhinkle** [comment]
## New Idea:

Enable Basic Auth for PEs that need to be locked down.

When enabling Basic Auth, the engineer provides a list of the people who should be able to access that PE in the form of a list of GitHub usernames. GitHub users who are part of the VA org (and who have whatever other permissions that are required like membership to a team, read/write access to a repo, etc.) are then emailed a short-lived password to access the PE.

In short, the same people who would be given access to the PE through OAuth would be given access through Basic Auth without:

- Any changes to `vets-website`
- Having to go through all the trouble of creating an OAuth app

**Comment by pjhill** [comment]
After talking to Identity team members we brainstormed the following three high level ideas for handling PE authn / authz --

1. Centralized GitHub OAuth solution that authenticates users before showing them a list of available preview environments. Then the authed user would have access to any preview environment.
2. Centralized basic auth solution that authenticates users with automatically generated basic auth credentials before showing them a list of available preview environments. Then the authed user would have access to any preview environment.
3. Instead of authenticating users after the fact, we authn/authz the user at the workflow execution time. If the user is a member of the organization then they are permitted to create a preview environment. If they are not then they do not get a preview environment.

**Comment by pjhill** [comment]
Here's a dump of my raw notes from the discussion --
* GitHub OAuth
   * There's more than DSVA
   * Veterans won’t be part of the VA org
   * GitHub OAuth with Sidekiq
      * If you go to the Sidekiq URL
      * You OAuth and it sets a cookie
* Send users to a centralized workflow where auth happens with GitHub OAuth
   * Redirect to a list of Pes
   * User clicks the one the want
   * Redirects to our FE container
   * The cookie for the preview.va.gov could propagate down to subdomains
   * Would not have granular control over 
* Can we restrict individuals who can create PEs through the PR mechanism?
   * We would not need authentication
   * Chelen approval?
   * Gary Fallon?
   * Ray Wang?
   * Bill Chapman?
   * ATO non-production systems are not part of ATO
* Push button deployments???
* Route53 additional cost
   * Pay for each domain
